### PR TITLE
Activity Classifier (Python) uses Tensorflow

### PR DESCRIPTION
**Pull request by shreyajain17**
This branch can be tested using the following command
(Also manually disable mps on Metal device)
```
import turicreate as tc
data = tc.SFrame('/Users/shreya_jain3/Downloads/hapt_data.sframe/')
model = tc.activity_classifier.create(data, target='activity', session_id='exp_id',max_iterations=10,validation_set=data[:20000],_advanced_parameters={'use_tensorflow' :True})
```
Dataset used - HAPT 

This shows training difference between tensorflow and MXNet.
```Using Tensorflow
2019-08-27 14:23:26.794216: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
+----------------+----------------+----------------+----------------+
| Iteration      | Train Accuracy | Train Loss     | Elapsed Time   |
+----------------+----------------+----------------+----------------+
| 1              | 0.584          | 1.103          | 1.5            |
| 2              | 0.772          | 0.647          | 2.8            |
| 3              | 0.834          | 0.474          | 4.1            |
| 4              | 0.865          | 0.378          | 5.5            |
| 5              | 0.882          | 0.337          | 6.7            |
| 6              | 0.895          | 0.292          | 7.9            |
| 7              | 0.912          | 0.254          | 9.1            |
| 8              | 0.915          | 0.228          | 10.4           |
| 9              | 0.922          | 0.212          | 11.7           |
| 10             | 0.926          | 0.199          | 12.9           |
+----------------+----------------+----------------+----------------+
Training complete
Total Time Spent: 12.9154s
+----------------+----------------+----------------+----------------+
| Iteration      | Train Accuracy | Train Loss     | Elapsed Time   |
+----------------+----------------+----------------+----------------+
| 1              | 0.596          | 1.012          | 1.5            |
| 2              | 0.786          | 0.572          | 3.0            |
| 3              | 0.843          | 0.416          | 4.4            |
| 4              | 0.870          | 0.340          | 5.8            |
| 5              | 0.888          | 0.291          | 7.2            |
| 6              | 0.902          | 0.257          | 8.6            |
| 7              | 0.910          | 0.234          | 9.9            |
| 8              | 0.917          | 0.214          | 11.2           |
| 9              | 0.924          | 0.190          | 12.5           |
| 10             | 0.931          | 0.174          | 13.8           |
+----------------+----------------+----------------+----------------+
Training complete
Total Time Spent: 13.8493s
```

I have also tested the working of this code with 

- [x] validation set (validation decreases but is higher than MXNet)
- [x] predict
- [x] evaluate
- [x] save 
- [x] load_model
- [x] load_model then predict, evaluate on data
- [x] export_coreml - I get the following warning `WARNING: Logging before flag parsing goes to stderr.
W0829 00:25:47.518141 4572663104 __init__.py:71] TensorFlow version 1.14.0 detected. Last version known to be fully compatible is 1.13.1 .`

`

### CODE REVIEW

File: scripts/requirements.txt [hunk 1]
```diff
@@ -26,3 +26,4 @@ pyOpenSSL==19.0.0
 ndg-httpsclient==0.5.1
 pyasn1==0.4.5
 hypothesis==4.24.3
+tensorflow==1.14.0
```
Comment (nickjong) [inline scripts/requirements.txt]: Do we want to make TensorFlow a formal requirement yet?

File: src/python/turicreate/toolkits/_tf_model.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+from __future__ import print_function as _
+from __future__ import division as _
+from __future__ import absolute_import as _
+import abc as _abc
+import six as _six
+
+@_six.add_metaclass(_abc.ABCMeta)
+class TensorFlowModel(object):
+	__metaclass__ = _abc.ABCMeta
```
Comment (nickjong) [inline src/python/turicreate/toolkits/_tf_model.py]: Is this still needed, given line 13?

File: src/python/turicreate/toolkits/_tf_model.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+
+	Make the graph
+	conv = tf.nn.conv1d(self.data, self.weights['conv'], ..) 
+	dense = tf.add(tf.matmul() + self.bias())
+	...
+
+
+    Make loss_op with the loss and train_op with the optimizer
```
Comment (nickjong) [inline src/python/turicreate/toolkits/_tf_model.py]: Inconsistent indentation here

File: src/python/turicreate/toolkits/_tf_model.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		NotImplementedError
+
+
+
+	"""
+	Train will do a forward and backward pass and update weights 
+    This accepts a dictionary that has feature/target as key and 
+    the numpy arrays as value corresponding to them respectively
+    It returns a dictionary of loss and output (probabilities)
+    This matches model backend train
```
[the remark this answers is not in the document]
Comment (nickjong) [reply src/python/turicreate/toolkits/_tf_model.py]: The keys the appear in the input and output dictionaries are toolkit-specific. We can't assert that it must always contain loss and probability output.

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py [hunk 5]
```diff
@@ -138,7 +139,9 @@ def create(dataset, session_id, target, features=None, prediction_window=100,
     from .._mps_utils import (use_mps as _use_mps,
                               mps_device_name as _mps_device_name,
                               ac_weights_mps_to_mxnet as _ac_weights_mps_to_mxnet)
-
+    from ._tf_model_architecture import ActivityTensorFlowModel, _fit_model_tf
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: Won't this always trigger an import of TensorFlow? We should only do this if the user has requested TensorFlow

It's important that this PR have no significant effect in master when `use_tensorflow` is false

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py [hunk 6]
```diff
@@ -199,13 +225,16 @@ def create(dataset, session_id, target, features=None, prediction_window=100,
             print('Using GPU to create model (CUDA)')
         elif num_mxnet_gpus > 1:
             print('Using {} GPUs to create model (CUDA)'.format(num_mxnet_gpus))
+        elif params['use_tensorflow']:
+            print('Using Tensorflow')
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: For consistency, how about "Using TensorFlow to create model"?

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py [hunk 7]
```diff
@@ -233,27 +262,82 @@ def create(dataset, session_id, target, features=None, prediction_window=100,
...
 
         log = _fit_model_mps(loss_model, data_iter, valid_iter, max_iterations, verbose)
-
     else:
-        # Train the model using Mxnet
-        log = _fit_model_mxnet(loss_model, data_iter, valid_iter,
-                         max_iterations, num_mxnet_gpus, verbose)
+        from .._mxnet import mxnet as _mx
+        # Initialize the weights in MXNet and then pass them in TensorFlow
+        dummy_data_iter = _SFrameSequenceIter(chunked_data, len(features),
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: Isn't this specific to `use_tensorflow`? Why are we doing this in the MXNet case too?

Could all this just be in `_fit_model_tf`?

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py [hunk 8]
```diff
@@ -233,27 +262,82 @@ def create(dataset, session_id, target, features=None, prediction_window=100,
...
     pred_model.bind(data_shapes=data_iter.provide_data, label_shapes=None,
                     for_training=False)
-
+    
     if use_mps:
         mps_params = loss_model.export()
         arg_params, aux_params = _ac_weights_mps_to_mxnet(mps_params, _net_params['lstm_h'])
+
+    elif params['use_tensorflow']:
+        # Copy the weights back in the MXNet format
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: Let's extract this code into a utility function appearing in your TF-specific file

File: src/python/turicreate/toolkits/activity_classifier/_mx_model_architecture.py [hunk 9]
```diff
@@ -104,9 +104,6 @@ def _lstm_layer(data, num_hidden, seq_len, dropout):
 def _fit_model_mxnet(model, data_iter, valid_iter, max_iterations, num_gpus, verbose):
     from time import time as _time
 
-    model.bind(data_shapes=data_iter.provide_data,
```
Comment (nickjong) [inline src/python/turicreate/toolkits/activity_classifier/_mx_model_architecture.py]: Let's avoid modifying the MXNet path. If necessary, you can pass the MXNet model to `_fit_model_tf` as part of the initialization of the TF model.

Note that we don't have to follow the same pattern as what was used for MXNet, if it doesn't make as much sense for us. Maybe just a function that takes all the needed inputs, constructs the TensorFlow model class, trains it, then returns the MXNet weights.

File: scripts/requirements.txt (hunk 1 above)
Comment (shreyajain17) [reply scripts/requirements.txt]: My bad. I don't want it to be a formal requirement yet.

File: src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py (hunk 7 above)
Comment (shreyajain17) [reply src/python/turicreate/toolkits/activity_classifier/_activity_classifier.py]: I did this to initialize the two with the same weights! I can move this since we are no longer comparing them with each other.
